designs/disruption-controls-drain-timeout.md

```diff
@@ -0,0 +1,78 @@
+# Disruption Controls - (in v1beta1+)
+
+## Motivation
+Users are requesting the ability to control how long karpenter will wait for a drain on an individual node to complete before forcefully evicting the remaining pods to decommission the node [#743](https://github.com/kubernetes-sigs/karpenter/issues/743). This supports two primary use cases.
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design piggybacks off the existing disruption-controls design, adding in a new `forceDrainAfter` field to the `v1beta/NodePool` `Disruption` spec. This field seems intuitively related to the other disruption controls that exist within the disruption block.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    forceDrainAfter: 24h || Never
```

Comment:
re: 
```
$ kubectl drain my-node --grace-period=0 --disable-eviction=true
```

thoughts on spec.disruption.gracePeriod? "The disruption grace period".

Reply:
The `grace-period=0` is actually not the necessary part of the drain command fwiw, the `disable-eviction` is the real lifter there. But I think that arg name is pretty unintuitive.

I'm not attached to a particular name and think something like `terminationGracePeriod` would be intuitive for users.

Reply:
TerminationGracePeriod I like. There's also `metadata.DeletionGracePeriodSeconds` in every Kubernetes object. In general, we've been using the term disruption, rather than deletion or termination, which is why I suggested `disruption.gracePeriod`. That said, this semantic is a bit awkward with the `Never` bit of our NillableDurations.

Reply:
Thoughts on having this be drainTimeout or terminationTimeout? We've been using the -After syntax to indicate a time when a node should be disrupted. I think it can apply here as well, but I think a timeout is a little more deliberate in the meaning and usage.

Alternatively, I like the GracePeriodSeconds + its derivatives.

Reply:
I think keeping `grace` is more clear because this isn't a timeout on termination necessarily nor does it need to be `drain` related? It is simply indicating a graceful period of time for termination was exceed and now we switch modes to ungracefully terminating the node and its pods? 

Grammatically I think we got forced into a corner if we have to stick with `Never`? Can it be something like `disruption.graceTimeout`

Reply:
We can always disallow `Never` and have 0 refer to an unlimited grace period like CAPI does if it helps avoid awkwardness.